# Python para Data Science: primeiros passos

## 01. Começando com Python

### Para saber mais: o que é Python?

**Python** é uma linguagem de programação de **alto nível**, **interpretada** e **orientada a objetos**. Foi criada no final de 1989 pelo programador holandês [Guido van Rossum](https://twitter.com/gvanrossum). O nome “Python” foi inspirado na série de televisão britânica *Monty Python's Flying Circus* e tem como símbolo a imagem mostrada abaixo:

![Python Logo](https://media0.giphy.com/media/v1.Y2lkPTc5MGI3NjExbG1tcmhmd2RjMXlndzVkazIzY2pqbjNlejFqYnE0aGFuNjd2emFyMSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/KAq5w47R9rmTuvWOWa/giphy.gif)

Python é utilizado em várias áreas como desenvolvimento de **aplicações web**, **ciência de dados**, **machine learning**, **automação de tarefas**, **análise de dados** e muitas outras. É uma linguagem de programação versátil que pode ser usada para criar praticamente qualquer tipo de programa.

Por ter uma **sintaxe simples e fácil de ler**, se torna uma boa escolha para iniciantes na programação. Além disso, Python é uma das linguagens mais usadas no mundo, possuindo uma grande comunidade ativa e muitas bibliotecas e frameworks disponíveis, o que facilita a realização de diversas tarefas.

Ele é compatível com diversos sistemas operacionais, como: Windows, Linux e macOS. Por ser uma linguagem popular e em constante evolução, Python é utilizado em muitas empresas e organizações em todo o mundo. Também é amplamente utilizado em aplicações científicas, devido à sua capacidade de lidar com grandes conjuntos de dados e com operações matemáticas complexas.

As informações e documentações sobre Python podem ser encontradas no site oficial [python.org](https://www.python.org/). Caso queira mais motivos para aprender essa linguagem, bem como se aprofundar nos conceitos destacados aqui, indico a leitura do artigo [O que é Python? - um guia completo para iniciar nessa linguagem de programação](https://www.alura.com.br/artigos/python?srsltid=AfmBOopEUbGpj_aK68p7YUTyp4NLNGC4-Sp6WxiHwHxZpFj3UCEcMqPi).

### Olá mundo!

In [1]:
print('Olá mundo!')

Olá mundo!


## 02. Manipulando dados no Python

### Variáveis

In [2]:
idade = 5

In [3]:
print(f"{idade=}")

idade=5


In [4]:
idade = 10
print (f"{idade=}")

idade=10


In [5]:
idade = 15
idade

15

In [6]:
nome = 'Gabriel'
nome

'Gabriel'

### Tipos de variáveis

| Classe | Descrição | Exemplo |
| :----: | :-------: | :-----: |
| int | Armazena números inteiros | 3, 100 ou -5 |
| float | Armazena números com ponto flutuante, ou seja, números que possuem uma parte fracionária ou reais | 3.14, 100.0 ou -0.5 |
| str | Armazena strings, ou seja, um texto ou sequências de caracteres. As strings são geralmente usadas para armazenar texto e são delimitadas po aspas simples ou duplas | "Olá!", 'Alura' ou "1234" |
| bool | Armazena valores lógicos, verdadeiro ou falso | True e False |

In [7]:
i = 5
type(i)

int

In [8]:
f = 9.8
type(f)

float

In [9]:
s = 'Cristoffer'
type(s)

str

In [10]:
b = True
type(b)

bool

In [11]:
nome_aluno = 'Fabricio Daniel'
idade_aluno = 15
media_aluno = 8.45
situacao_aprovado = True
print(nome_aluno, idade_aluno, media_aluno, situacao_aprovado)

Fabricio Daniel 15 8.45 True


### Variáveis numéricas

| Cargo | Quantidade | Salário |
| :---- | :--------- | :------ |
| Segurança | 5 | 3000 |
| Docente | 16 | 6000 |
| Diretoria | 1 | 12500 |

Precisamos trabalhar com esses dados fornecendo:
- A quantidade total de empregados;
- A diferença entre o salário mais baixo e mais alto; e
- A média ponderada da faixa salarial da escola

In [12]:
q_seguranca = 5
s_seguranca = 3000

q_docente = 16
s_docente = 6000

q_diretoria = 1
s_diretoria = 12500

In [15]:
total_empregados = q_seguranca + q_docente + q_diretoria
total_empregados

22

In [16]:
diferenca_salario = s_diretoria - s_seguranca
diferenca_salario

9500

In [17]:
media_salario = (q_seguranca * s_seguranca + q_docente * s_docente + q_diretoria * s_diretoria) / total_empregados
media_salario

5613.636363636364

### Strings

In [18]:
s1 = 'Alura'
s2 = "Alura"
print(type(s1), type(s2))

<class 'str'> <class 'str'>


In [19]:
dir(s1)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'removeprefix',
 'removesuffix',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'stri

In [20]:
texto = '  Geovana Alessandra dias Sanyos '

In [21]:
[print(p.capitalize().strip(), end=' ') for p in texto.replace('y', 't').split()]

Geovana Alessandra Dias Santos 

[None, None, None, None]

In [23]:
texto = ' '.join([p.capitalize().strip() for p in texto.replace('y', 't').split()])
texto

'Geovana Alessandra Dias Santos'

### Para saber mais: tabela Unicode

Python é uma linguagem de programação caracterizada por usar o padrão unicode de caracteres. Esse padrão permite que pessoas desenvolvedoras de todo o mundo consigam trabalhar com textos e caracteres em diferentes idiomas de forma muito fácil, sem precisar se preocupar com problemas de compatibilidade.

Isso só foi possível graças ao desenvolvimento da **tabela Unicode**, uma tabela de codificação de caracteres, que associa códigos numéricos a caracteres específicos. O objetivo de sua criação foi incluir os caracteres existentes em todas as línguas e sistemas escritos do mundo. Até a versão 15.0, o padrão já agrupa mais de 140 mil caracteres, incluindo letras, números, símbolos e emojis. Podemos ver as estatísticas de cada versão e a quantidade de caracteres por versão nesse [link](https://www.unicode.org/versions/stats/).

Além disso, ela veio como uma alternativa às outras tabelas de caracteres criadas como a ASCII, que só suporta caracteres do alfabeto inglês e alguns símbolos. Por conta disso, ela é muito importante, já que permite a representação de caracteres de diferentes idiomas (como o ç no Português, o ñ no Espanhol ou o П em Russo) e scripts de forma consistente e independentemente do dispositivo ou plataforma, o que é fundamental para a globalização e representação correta de textos em diferentes línguas.

Vale lembrar que a tabela Unicode é constantemente atualizada e ampliadam incluindo novos caracteres e scripts. Isso garante que ela continue sendo relevante e útil para suportar a representação de todos os idiomas e scripts do mundo. Você pode acompanhar as atualizações no site [unicode.org](https://home.unicode.org/) e acessar as codificações de cada caractere desse padrão na [wikipedia](https://en.wikipedia.org/wiki/List_of_Unicode_characters) ou por [PDFs oficiais](http://unicode.org/charts/).

Podemos imprimir um elemento da tabela Unicode em nosso notebook através da função `chr`, que é uma função que retorna uma string com o caractere especificado. Sua sintaxe de uso é dada da seguinte forma: `chr(numero_do_caractere)`. Como exemplo, vamos imprimir o caractere “@” com `chr()`. Segundo a [tabela](https://en.wikipedia.org/wiki/List_of_Unicode_characters#:~:text=U%2B0040-,%40,-64), sabemos que o `@` corresponde ao número `64`, então reproduzimos:

```python
chr(64)
```
> '@'

Várias outras combinações de números podem ser testadas e observadas nas saídas. Sinta-se livre para testar várias combinações e até mesmo criar frases!

Também é possível **concatenar strings** com o operador `+`. Sabendo disso, se nós unirmos vários caracteres, conseguimos obter uma palavra. Como exemplo, vamos construir a palavra `Olá` (`'O' + 'l' + 'á'`):

```python
chr(79) + chr(108) + chr(225)
```
> 'Olá'

### Coletando dados

In [25]:
nome = input('Escreva seu nome: ')

Escreva seu nome:  Cristoffer


In [26]:
nome

'Cristoffer'

In [27]:
ano_entrada = input('Escreva o ano de ingresso do(a) estudante: ')

Escreva o ano de ingresso do(a) estudante:  2023


In [28]:
type(ano_entrada)

str

In [29]:
ano_entrada = int(input('Escreva o ano de ingresso do(a) estudante: '))

Escreva o ano de ingresso do(a) estudante:  2023


In [30]:
type(ano_entrada)

int

In [31]:
nota_entrada = float(input('Digite a nota do teste de ingresso: '))
print(f"Ano de entrada {ano_entrada} - nota do teste de ingresso {nota_entrada}")

Digite a nota do teste de ingresso:  8.5


Ano de entrada 2023 - nota do teste de ingresso 8.5


### Para saber mais: formatações do print()

Conseguimos visualizar o resultado de variáveis dentro de strings, como como imprimir o texto final em um print. Durante a aula, aprendemos a utilizar a formatação ***f-string*** (ou formatação de string), na qual colocamos um `f` antes da criação da strinf e as variáveis entre chaves `{}`. Exemplo:

```python
nome = 'Ana Maria'
idade = 17
print(f"O nome da aluna é {nome} e sua idade é {idade} anos.")
```
> 'O nome da aluna é Ana Maria e sua idade é 17 anos'

Mas existem outras formatações, como o uso do operador de formatação de string ou da função `.format()`.

#### Operador de formatação

Esse operador de formatação permite a inserção de váriáveis em pontos específicos na string com o operador `%`. Esse operador funciona como um marcador, informando onde o valor da variável vai ser exposto na string.

O `%` precisa ser acompanhado de uma palavra-chave para cada tipo de variável que se deseja adicionar. Seguindo a tabela abaixo:

| Tipo de variável | Palavra-chave |
| :--------------- | :------------ |
| string | `%s` |
| int | `%d` |
| float | `%f` |
| chr | `%c` |

Desse modo, para inserirmos uma variável, podemos adicionar o operador pela string no ponto desejado. Após o fim da string, adicionamos novamente o `%`, mas agora especificando a variável entre parênteses. Podemos observar essa estrutura no exemplo abaixo:

```python
nome_aluno = 'Fabricio Daniel'
print('Nome do aluno: %s' %(nome_aluno))
```
> 'Nome do aluno: Fabricio Daniel'

Caso tenha mais de uma variável, devemos ordená-las conforme seu surgimento no texto e separá-las por vírgula.

```python
nome_aluno = 'Fabricio Daniel'
idade_aluno = 15
media_aluno = 8.45
print('Nome do aluno é %s, ele tem %d anos e sua média é %f.' %(nome_aluno, idade_aluno, media_aluno))
```
> 'Nome do aluno é Farbricio Daniel, ele tem 15 anos e sua média é 8.4500000.'

Quando trabalhamos com pontos flutuantes (`float`), podemos determinar a quantidade de casas decimais após a vírgula usando a sintaxe `%.xf`, na qual o `x` é o número de casas desejadas. Utilizando as mesmas variáveis do exemplo anterior, o código com `%.xf` fica da seguinte maneira:

```python
print('Nome do aluno é %s, ele tem %d anos e sua média é %.2f.' %(nome_aluno, idade_aluno, media_aluno))
```
> 'Nome do aluno é Farbricio Daniel, ele tem 15 anos e sua média é 8.45.'

Uma observação: os operadores de formatação de strings com `%` não funcionam diretamente com valores booleanos. Uma maneira de lidar com isso é convertendo o valor booleano para uma string antes de usá-lo na formatação com a função `str()`. Por exemplo:

```python
x = True
print('Valor de x: %s' % str(x))
```

Isso não acontece com a formatação *f-string* ou *.format*.

#### `format`

É possível também usar o método `format()` para fazer a formatação de strings. Ele é mais flexível e permite passar as variáveis diretamente dentro da string, sem a necessidade dos operadores `%`. Seus marcadores são apenas as `{}`. Exemplo:

```python
nome_aluno = 'Fabricio Daniel'
print('Nome do aluno {}'.format(nome_aluno))
```
> 'Nome do aluno Fabricio Daniel'

Do mesmo modo que é feito com o operador, conseguimos aplicar com mais variáveis:

```python
nome_aluno = 'Fabricio Daniel'
idade_aluno = 15
media_aluno = 8.45
print('Nome do aluno é {}, ele tem {} anos e sua média é {}'.format(nome_aluno, idade_aluno, media_aluno))
```
> 'Nome do aluno é Fabricio Daniel, ele tem 15 anos e sua média é 8.45.'

Note que, com o `format`, não temos o problema das casas decimais do ponto flutuante. Esse problema também não é presente no formato f-string.

Em resumo, cada uma dessas formas tem suas vantagens e desvantagens. Indico utilizar a forma f-string, pois é mais legível e fácil de usar. Mesmo assim, cada pessoa desenvolvedora pode escolher a forma que achar mais apropriada.

#### Caracteres especiais

Além da formatação de inserção de variáveis dentro de uma string, também existem os **caracteres especiais**. Eles são usados parar representar ações especiais ou caracteres que não podem ser digitados diretamente, como o Enter e a tabulação. Vamos conhecer alguns deles? Busque reproduzir todos os exemplos e visualizar o resultado final.

- `\n` é o caractere de nova linha e é usado para pular uma linha no texto (função do Enter). Exemplo:

```python
print("Estudar é um esforço constante,\nÉ como cultivar uma planta,\nPrecisamos de dedicação e paciência,\nPara ver o fruto amadurecer.")
```

- `\t` é o caractere de tabulação usado para adicionar um espaço de tabulação no texto. Exemplo:

```python
print('Quantidade\tQualidade\n5 amostras\tAlta\n3 amostras\tBaixa')
```

- `\\` é usado para imprimir uma única barra invertida. Caso não seja usada a dupla barra invertida, o código poderá resultar em erro ou em um resultado inesperado, pois o Python considera a `\` um chamado para um caractere especial. Usamos esta sintaxe para garantir que não ocorram erros. Exemplo:

```python
print("Caminho do arquivo: C:\\arquivos\\documento.csv")
```

- `\"` é usado para imprimir aspas duplas quando estamos trabalhando com uma string criada a partir de aspas duplas `"`. Porém, isso não é necessário caso seja uma string criada por aspas simples `'`. Como exemplo, temos:

```python
print("Ouvi uma vez \"Os frutos do conhecimento são os mais doces e duradouros de todos.\"")
```

- `\'` é usado para imprimir aspas simples quando estamos trabalhando com uma string criada a partir de aspas simples `'`. Caso seja uma string criada por aspas dulpas `"`, isso não é necessário. Exemplo:

```python
print('Minha professora uma vez disse \'Estudar é a chave do sucesso.\' ')
```

## 03. Estruturas condicionais

### Conhecendo o if e o else

In [32]:
if 2 < 7:
    print('condição verdadeira')

condição verdadeira


### Para saber mais: operadores relacionais

Uma condição nada mais é que uma comparação entre dados, a partir da qual nó podemos obter o resultado verdadeiro ou falso de uma condicional. A comparação pode ser feita por operadores relacionais que tem a missão de comparar valores e determinar se uma expressão é verdadeira ou falsa. Vamos conhecer os comparadores lógicos e como podemos usá-los.

#### *Maior que* (`>`)

Retorna verdadeiro se o valor à esquerda do símbolo for **maior** que o da direita. Sua sintaxe, `valor_1 > valor_2`, representa uma comparação que só retornará verdadeiro se valor_1 for maior que valor_2. Com esse comparador lógico, podemos verificar se uma pessoa é mais velha que outra ao comparar suas idades como mostrado no exemplo a seguir:

```python
idade_maria = int(input('Digite a idade da Maria: '))
idade_beatriz = int(input('Digite a idade da Beatriz: '))

if idade_maria > idade_beatriz:
  print('Maria é mais velha que Beatriz.')
```
> Digite a idade da Maria: 12
> Digite a idade da Beatriz: 10
> 'Maria é mais velha que Beatriz'

O comparador maior que (`>`) retorna um valor falso caso os valores comparados sejam iguais ou o valor à esquerda do símbolo for inferior ao da direita.

#### *Menor que* (`<`)

Retorna verdadeiro se o valor à esquerda do símbolo for **menor** que o da direita e sua sintaxe é a seguinte: `valor_1 < valor_2`. Essa comparação só retornará verdadeiro se valor_1 for menor que valor_2. Com esse operador relacional, também podemo verificar se uma pessoa é mais velha que outra comparando suas idades:

```python
idade_maria = int(input('Digite a idade da Maria: '))
idade_beatriz = int(input('Digite a idade da Beatriz: '))

if idade_maria < idade_beatriz:
  print('Maria é mais nova que Beatriz.')
```
> Digite a idade da Maria: 30
> Digite a idade da Beatriz: 34
> 'Maria é mais nova que Beatriz'

Esse comparador também retorna um valor falso caso os comparados sejam iguais ou se o valor à esquerda do símbolo for superior ao da direita.

#### *Maior ou igual* (`>=`)

Retorna verdadeiro se o valor à esquerda do símbolo for **maior ou igual** ao valor da direita. Sua sintaxe é a seguinte: `valor_1 >= valor_2`. Essa comparação só retornará verdadeiro se valor_1 for maior ou igual ao valor_2. Com esse operador relacional podemos, por exemplo, verificar se uma empresa tem uma quantidade maior ou igual à outra:

```python
empregados_empresa_1 = int(input('Digite a quantidade de empregados da empresa 1: '))
empregados_empresa_2 = int(input('Digite a quantidade de empregados da empresa 2: '))

if empregados_empresa_1 >= empregados_empresa_2:
  print('Empresa 1 tem uma quantidade de empregados maior ou igual à empresa 2.')
```
> Digite a quantidade de empregados da empresa 1: 300
> Digite a quantidade de empregados da empresa 2: 150
> 'Empresa 1 tem uma quantidade de empregados maior ou igual à empresa 2.'

#### *Menor ou igual* (`<=`)

Retorna verdadeiro se o valor à esquerda do símbolo for **menor ou igual** ao valor à direita. Sua sintaxe é a seguinte: `valor_1 <= valor_2`. Essa comparação só retornará verdadeiro se valor_1 for menor ou igual ao valor_2. Com esse operador relacional, verificamos se uma empresa tem uma quantidade menor ou igual à outra, como mostrado a seguir:

```python
empregados_empresa_1 = int(input('Digite a quantidade de empregados da empresa 1: '))
empregados_empresa_2 = int(input('Digite a quantidade de empregados da empresa 2: '))

if empregados_empresa_1 <= empregados_empresa_2:
  print('Empresa 1 tem uma quantidade de empregados menor ou igual à empresa 2.')
```
> Digite a quantidade de empregados da empresa 1: 200
> Digite a quantidade de empregados da empresa 2: 200
> 'Empresa 1 tem uma quantidade de empregados menor ou igual à empresa 2.'

#### *Igual a* (`==`)

Retorna verdadeiro se o valor à esquerda do símbolo for **igual** ao valor à direita. Sua sintaxe, `valor_1 == valor_2`, representa uma comparação que só retornará verdadeiro se valor_1 for igual ao valor_2. Com esse operador relacional, podemos verificar, por exemplo, se dois livros têm o mesmo nome através de uma comparação de strings:

```python
livro_1 = input('Digite o título do 1º livro: ')
livro_2 = input('Digite o título do 2º livro: ')

if livro_1 == livro_2:
  print('Os livros têm o mesmo título')
```
> Digite o título do 1º livro: Verso e Código: A poética da programação
> Digite o título do 2º livro: Verso e Código: A poética da programação
> 'Os livros têm o mesmo título'

Também é possível fazer essa comparação com valores numéricos.

#### *Diferente de* (`!=`)

Retorna verdadeiro se o valor à esquerda do símbolo for **diferente** do valor à direita. Sua sintaxe, `valor_1 != valor_2`, representa uma comparação que só retornará verdadeiro se valor_1 for diferente do valor_2. Esse operador é o inverso do `==`. Com ele, podemos verificar se dois livros têm títulos diferentes através de uma comparação de strings:

```python
livro_1 = input('Digite o título do 1º livro: ')
livro_2 = input('Digite o título do 2º livro: ')

if livro_1 != livro_2:
  print('Os livros têm títulos diferentes')
```
> Digite o título do 1º livro: Verso e Código: A poética da programação
> Digite o título do 2º livro: Programar com versos: A poesia do código
> 'Os livros têm títulos diferentes'

Também é possível fazer essa comparação com valores numéricos.

### Utilizando if e else

In [34]:
media = float(input('Digite a média: '))

if media >= 6.0:
    print('Aprovado(a)')
else:
    print('Reprovado(a)')

Digite a média:  5.0


Reprovado(a)


In [36]:
media = float(input('Digite a média: '))

if media >= 6.0:
    print('Aprovado(a)')
elif 6.0 > media >= 4.0:
    print('Recuperação')
else:
    print('Reprovado(a)')

Digite a média:  5.0


Recuperação


In [37]:
media = float(input('Digite a média: '))

if media >= 6.0:
    print('Aprovado(a)')
if 6.0 > media >= 4.0:
    print('Recuperação')
else:
    print('Reprovado(a)')

Digite a média:  7.0


Aprovado(a)
Reprovado(a)


### Outras alternativas com elif

In [38]:
media = float(input('Digite a média: '))

if media >= 6.0:
    print('Aprovado(a)')
elif 6.0 > media >= 4.0:
    print('Recuperação')
else:
    print('Reprovado(a)')

Digite a média:  5.0


Recuperação


### Operadores lógicos

In [39]:
t1 = t2 = True
f1 = f2 = False

In [42]:
if t1 and t2:
    print('expressão verdadeira')
else:
    print('expressão falsa')

expressão verdadeira


In [43]:
if t1 and f2:
    print('expressão verdadeira')
else:
    print('expressão falsa')

expressão falsa


In [44]:
if f1 or f2:
    print('expressão verdadeira')
else:
    print('expressão falsa')

expressão falsa


In [45]:
if t1 or f2:
    print('expressão verdadeira')
else:
    print('expressão falsa')

expressão verdadeira


In [46]:
if not t1:
    print('expressão verdadeira')
else:
    print('expressão falsa')

expressão falsa


In [47]:
if not f1:
    print('expressão verdadeira')
else:
    print('expressão falsa')

expressão verdadeira


In [49]:
lista = 'José da Silva, Maria Oliveira, Pedro Martins, Ana Souza, Carlos Rodrigues, Juliana Santos, Bruno Gomes, Beatriz Costa, Felipe Almeida, Mariana Fernandes, João Pinto, Luísa Nascimento, Gabriel Souza, Manuela Santos, Thiago Oliveira, Sofia Ferreira, Rafael Albuquerque, Isabella Gomes, Bruno Costa, Maria Martins, Rafaela Souza, Matheus Fernandes, Luísa Almeida, Beatriz Pinto, Mariana Rodrigues, Gabriel Nascimento, João Ferreira, Maria Albuquerque, Felipe Oliveira'
lista

'José da Silva, Maria Oliveira, Pedro Martins, Ana Souza, Carlos Rodrigues, Juliana Santos, Bruno Gomes, Beatriz Costa, Felipe Almeida, Mariana Fernandes, João Pinto, Luísa Nascimento, Gabriel Souza, Manuela Santos, Thiago Oliveira, Sofia Ferreira, Rafael Albuquerque, Isabella Gomes, Bruno Costa, Maria Martins, Rafaela Souza, Matheus Fernandes, Luísa Almeida, Beatriz Pinto, Mariana Rodrigues, Gabriel Nascimento, João Ferreira, Maria Albuquerque, Felipe Oliveira'

In [51]:
nome_1 = 'Mariana Rodrigues'
nome_2 = 'Marcelo Nogueira'

In [52]:
if nome_1 in lista:
    print(f"{nome_1} está na lista")
else:
    print(f"{nome_1} não está na lista")

Mariana Rodrigues está na lista


In [53]:
if nome_2 in lista:
    print(f"{nome_2} está na lista")
else:
    print(f"{nome_2} não está na lista")

Marcelo Nogueira não está na lista


### Para saber mais: tabela verdade

Quando desejamos criar uma expressão lógica com operadores lógicos, precisamos saber como funciona a **tabela verdade** de cada um deles para conseguirmos construir um código que resulte no que esperamos. Essas tabelas ajudam a entender e a analisar expressões lógicas, como as encontradas em algoritmos de programação e para verificar a validade de uma dada expressão lógica. A tabela auxilis na verificação dos resultados para todas as possíveis combinações de valores lógicos de uma dada expressão lógica.

Vamos conhecer a tabela verdade dos operadores lógicos `and`, `or` e `not`.

#### Operador `and`

A tabela verdade do operador `and` nos mostra que a saída somente retorna Verdade (`True`) se ambos os operandos são Verdade. Caso contrário, retorna Falso (`False`), como mostrado na tabela abaixo:

| `operando_1` | `operando_2` | `operando_1 and operando_2` |
| :----------- | :----------- | :-------------------------- |
| False | False | False |
| False | True | False |
| True | False | False |
| True | True | True |

#### Operador `or`

A tabela verdade do operador `or` nos mostra que a saída somente retorna Falso (`False`) se ambos os operandos são Falsos, caso contrário, retorna Verdade (`True`). Ou seja, será retornado `True` se pelo menos um dos operandos for `True`, como mostrado abaixo:

| `operando_1` | `operando_2` | `operando_1 and operando_2` |
| :----------- | :----------- | :-------------------------- |
| False | False | False |
| False | True | True |
| True | False | True |
| True | True | True |

O operador `not` tem uma tabela verdade menor, uma vez que ele apenas inverte o valor do operando. Se o operando for Verdadeiro, retorna Falso; se for Falso, retorna Verdadeiro.

#### Operador `not`

| `operando` | `not operando` |
| :--------- | :------------- |
| True | False |
| False | True |

## 04. Estruturas de repetição

### Laço while

In [55]:
contador = 1
while contador <= 10:
    print(contador)
    contador += 1

1
2
3
4
5
6
7
8
9
10


### Para saber mais: operadores de atribuição

**Operadores de atribuição** permitem atribuir um dado a uma variável. Já trabalhamos bastante com o principal deles no Python, o `=`, que faz a atribuição literal de um valor a uma dada variável, como `idade = 20`, `escolaridade = 'superior'`, etc.

Também conhecemos o operador de soma `+=` que adiciona um valor especificado na variável, como consta no exemplo a seguir:

```python
preco = 2.00
preco += 3
print(preco)
```
> 5.0

Além desses, existem outros operadores de atribuição que permitem alterar os valores das variáveis. As informações de operadores, descrição e exemplo podem ser encontradas na tabela abaixo.

| Operador | Descrição | Exemplo |
| :------- | :-------: | ------: |
| `-=` | Subtrai um valor na variável | `preco -= 5` |
| `*=` | Multiplica um valor pela variável | `preco *= 3` |
| `/=` | Divide a variável por um valor | `preco /= 2` |
| `//=` | Realiza a divisão inteira da variável por um valor | `preco //= 6` |
| `%=` | Calcula o resto da divisão do valor na variável e atribui o resultado à variável | `preco %= 5` |

### Laço for

In [56]:
for i in range(1, 11):
    print(i)

1
2
3
4
5
6
7
8
9
10


### Para saber mais: comandos de controle

Ao trabalharmos com laços, conseguimos controlar o fluxo de execuções dentro do bloco de código, permitindo manipular a execução dos laços. O `continue` e o `break` são os comandos de controle que podemos executar com `for` e `while`.

O `continue` interrompe a iteração atual do laço e salta para a próxima, ou seja, retorna ao início do código. Como exemplo, temos um código de contagem de `1` a `5` com `for`:

```python
for i in range(1, 6):
    if i == 4:
        continue
    print(i)
```

Esse código imprime todos os números de 1 a 5, exceto o 4. Quando o valor de `i` for 4, o `continue` salta para a próxima iteração, ignorando o comando `print` abaixo da condicional na iteração atual.

Já o `break` interrompe a execução do laço completamente, saindo do bloco de código. Vamos usar o mesmo exemplo de contagem, mas agora colocaremos o `break`:

```python
for i in range(1, 6):
    if i == 4:
        break
    print(i)
```

Nesse caso, o código imprime todos os números de 1 a 3. Quando o valor de `i` for 4, o `break` interrompe a execução do laço e sai dele, ignorando qualquer outra iteração que esteja dentro da estrutura.

Em resumo, o `continue` permite pular uma iteração específica do laço, enquanto o `break` permite finalizar o laço e continuar a execução dos comandos fora dele, como mostrado abaixo:

![`continue` and `break`](https://cdn3.gnarususercontent.com.br/2922-python-data-science/aula4/aula4-img1.png)

## 05. Estruturas de dados

### Listas

In [57]:
lista = ['Fabricio Daniel', 9.5, 9.0, 8.0, True]
lista

['Fabricio Daniel', 9.5, 9.0, 8.0, True]

                    [0]         [1]  [2]  [3]  [4]
```python
lista = ['Fabricio Daniel', 9.5, 9.0, 8.0, True]
```

                    [-5]        [-4] [-3] [-2] [-1]
```python
lista = ['Fabricio Daniel', 9.5, 9.0, 8.0, True]
```

In [59]:
for elemento in lista:
    print(elemento)

Fabricio Daniel
9.5
9.0
8.0
True


In [60]:
lista[3] += 2.0
lista

['Fabricio Daniel', 9.5, 9.0, 10.0, True]

In [63]:
media = (lista[1] + lista[2] + lista[3]) / 3
media

9.5

### Para saber mais: String - uma sequência de caracteres

Como visto em aulas passadas, string é um tipo de dado que corresponde a um dado textual. Quando criamos uma string, estamos agrupando vários caracteres (números, letras e até símbolos) e cada um deles tem seus índices. Como exemplo, vamos criar uma string com o nome *Python*:

```python
linguagem = 'Python'
```

Cada caractere da string `linguagem` pode ser acessado através de seu índice que inicia em `0` e vai até a quantidade de caracteres da string menos 1, incluindo os índices negativos. Conseguimos acessar eles do mesmo modo que fazemos nas listas:

```python
print(linguagem[0], linguagem[1], linguagem[2], linguagem[-3], linguagem[-2], linguagem[-1])
```
> P y t h o n

No entanto, **os índices servem apenas para acessar os dados**, não sendo possível alterar o caractere presente em um índice definido através de uma simples atribuição, como nas listas. Por exemplo, o código `linguagem[0] = 'p'` gera um erro na compilação.

Com isso, podemos pensar que a string é uma estrutura de dados da mesma forma que as listas, certo? Na verdade, não. A string é uma sequência de caracteres (letras, números, símbolos, etc.) representada por uma única variável. Já uma estrutura de dados armazena uma coleção de itens (podendo ser de tipos diferentes) em uma única variável.

Mesmo assim, é possível transformar uma string em uma lista através do método `split()`. Esse método separa a string em uma lista de strings, usando um delimitador especificado entre parênteses. Esse delimitador precisa ser uma string. Como exemplo, vamos transformar a string em lista separando-a cada vez que aparece o ponto de interrogação `?`:

```python
duvida = 'Quem veio antes? O ovo? Ou foi a serpente?'
lista_palavras = duvida.split('?')
print(lista_palavras)
```
> ['Quem veio antes', 'O ovo', 'Ou foi a serpente', '']

O delimitador não aparece na separação. Caso ele não seja definido, a string será segregada por todos os espaços em branco no texto.

```python
duvida = 'Quem veio antes? O ovo? Ou foi a serpente?'
lista_palavras = duvida.split()
print(lista_palavras)
```
> ['Quem', 'veio', 'antes?', 'O', 'ovo?', 'Ou', 'foi', 'a', 'serpente?']

O contrário também acontece, pois podemos transformar uma lista em uma string através do método `join()`. Para usar a função, precisamos definir o caractere unificador que será utilizado para unir os elementos da lista, formando a string. Em seguida, usamos o método `join` passando a lista como argumento. Vamos fazer um exemplo com uma lista contendo o resultado de algumas misturas de cores primárias em tintas:

```python
misturas = ['Tintas: vermelho, azul e amarelo',
            'Verde: mistura de azul e amarelo',
            'Laranja: mistura de vermelho e amarelo',
            'Roxo: mistura de vermelho e azul']
unificador = '. '
string_misturas = unificador.join(misturas)
print(string_misturas)
```
> Tintas: vermelho, azul e amarelo. Verde: mistura de azul e amarelo. Laranja: mistura de vermelho e amarelo. Roxo: mistura de vermelho e azul

### Manipulaçao de listas

##### Quantidade de elementos
Usamos a função `len()` para descobrirmos a quantidade de elementos de um conjunto

In [65]:
len(lista)

5

##### Partição
A partição de listas por indexação em Python é uma técnica muito útil para selecionar um subconjunto de elementos de uma lista. Ela é feita usando a sintaxe `lista[início:fim]`, onde `inicio` é o índice do primeiro elemento a ser incluído na partição e `fim` é o índice do primeiro elemento a ser excluído da partição.

In [67]:
lista[1:4]

[9.5, 9.0, 10.0]

In [68]:
lista[3:]

[10.0, True]

##### `append()`

Adiciona um elemento ao final da lista.

In [69]:
lista.append(media)
lista

['Fabricio Daniel', 9.5, 9.0, 10.0, True, 9.5]

##### `extend()`
Adiciona vários elementos ao final da lista.
Adicionaremos as notas `[10.0, 8.0, 9.0]` na lista do Fabricio Daniel

In [70]:
lista.extend([10.0, 8.0, 9.0])
lista

['Fabricio Daniel', 9.5, 9.0, 10.0, True, 9.5, 10.0, 8.0, 9.0]

*Isso não é possível ser feito com o* `append`.

In [71]:
lista.append([10.0, 8.0, 9.0])
lista

['Fabricio Daniel',
 9.5,
 9.0,
 10.0,
 True,
 9.5,
 10.0,
 8.0,
 9.0,
 [10.0, 8.0, 9.0]]

##### `remove()`
Remove um elemento específico da lista.

In [72]:
lista.remove([10.0, 8.0, 9.0])
lista

['Fabricio Daniel', 9.5, 9.0, 10.0, True, 9.5, 10.0, 8.0, 9.0]

### Para saber mais: outras manipulações para listas

Estudamos alguns métodos em aula e, além deles, podemos usar outros para manipular listas em Python. Vamos conhecê-los e ver exemplos de sua construção. Para todos os casos usaremos a lista `raca_caes` a seguir:

```python
raca_caes = ['Labrador Retriever',
             'Bulldog Francês',
             'Pastor Alemão',
             'Poodle']
```

O primeiro método é o **`insert()`** que insere o elemento em uma determinada posição da lista. Para isso, seguimos a sintaxe `lista.insert(índice, elemento)`, na qual `lista` é a lista que vai receber o novo elemento, `índice` é o índice onde será inserido o novo elemento e `elemento` é o novo elemento a ser inserido.

```python
raca_caes.insert(1, 'Golden Retriever')
raca_caes
```
> ['Labrador Retriever', 'Golden Retriever', 'Bulldog Francês', 'Pastor Alemão', 'Poodle']

Seguindo esse pensamento, a estrutura `lista.insert(len(lista), elemento)` corresponde ao uso do método `append`, como vimos no vídeo anterior intitulado [Manipulação de listas](https://cursos.alura.com.br/course/python-data-science-primeiros-passos/task/122399).

O método **`pop()`** remove o elemento de uma determinada posição da lista e o retorna como saída na execução do método. Nele só precisamos especificar, nos parênteses, qual o índice do elemento que queremos remover e ele será apagado da lista. Portanto, vamos remover a raça `Golden Retriever` que adicionamos no método anterior.

```python
raca_caes.pop(1)
```
> 'Golden Retriever'

O método **`index()`** retorna o índice de um elemento específico na lista. Para isso, especificamos o elemento entre parênteses. Para identificar o índice da raça *Pastor Alemão* na lista, fazemos:

```python
raca_caes.index('Pastor Alemão')
```
> 2

O método **`sort()`** organiza os elementos da lista em ordem crescente ou decrescente. Caso sejam palavras, a ordem é dada por sequência alfabética ou inversa à ela. Para ordenar os valores, podemos apenas declarar o método `sort` que a lista será organizada em ordem. Para ordenar alfabéticamente a lista `raca_caes` temos o seguinte código:

```python
raca_caes.sort()
raca_caes
```
> ['Bulldog Francês', 'Labrador Retriever', 'Pastor Alemão', 'Poodle']

Para mais opções e descrições de métodos, acesse a [documentação do Python](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists).

### Dicionário

In [73]:
dicionario = {'chave_1': 1,
              'chave_2': 2}
dicionario

{'chave_1': 1, 'chave_2': 2}

In [74]:
cadastro = {'matricula': 2000168933,
            'dia_cadastro': 25,
            'mes_cadastro': 10,
            'turma': '2E'}
cadastro

{'matricula': 2000168933,
 'dia_cadastro': 25,
 'mes_cadastro': 10,
 'turma': '2E'}

In [75]:
cadastro.get('matricula', None)

2000168933

In [76]:
cadastro['turma'] = '2G'
cadastro

{'matricula': 2000168933,
 'dia_cadastro': 25,
 'mes_cadastro': 10,
 'turma': '2G'}

In [77]:
cadastro['modalidade'] = 'EAD'
cadastro

{'matricula': 2000168933,
 'dia_cadastro': 25,
 'mes_cadastro': 10,
 'turma': '2G',
 'modalidade': 'EAD'}

### Aprofundando em dicionários

##### `pop()`
Remove um item de um dicionário e o retorna.

In [78]:
cadastro.pop('turma')

'2G'

In [79]:
cadastro

{'matricula': 2000168933,
 'dia_cadastro': 25,
 'mes_cadastro': 10,
 'modalidade': 'EAD'}

##### `items()`
Retorna uma lista de pares de chave-valor do dicionário.

In [80]:
cadastro.items()

dict_items([('matricula', 2000168933), ('dia_cadastro', 25), ('mes_cadastro', 10), ('modalidade', 'EAD')])

##### `keys()`
Retorna uma lista das chaves do dicionário.

In [81]:
cadastro.keys()

dict_keys(['matricula', 'dia_cadastro', 'mes_cadastro', 'modalidade'])

##### `values()`
Retorna uma lista dos valores do dicionário.

In [82]:
cadastro.values()

dict_values([2000168933, 25, 10, 'EAD'])

In [83]:
for k in cadastro.keys():
    print(cadastro[k])

2000168933
25
10
EAD


In [84]:
for v in cadastro.values():
    print(v)

2000168933
25
10
EAD


In [85]:
for k, v in cadastro.items():
    print(k, v)

matricula 2000168933
dia_cadastro 25
mes_cadastro 10
modalidade EAD


### Para saber mais: listas em dicionários

Podemos associar estruturas de dados a outras estruturas de dados, como acontece com as listas dentro de dicionários. Nesse caso, as listas podem ser armazenadas nos valores de um dicionário de forma que cada chave pode ter uma lista associada a ela. Isso é útil quando precisamos armazenar vários valores relacionados a uma única chave. Por exemplo: podemos construir um conjunto de dados de uma loja de forma que contenha uma chave com os nomes de cada produto e outra chave contendo os preços correspondentes, como consta no código a seguir:

```python
loja = {'nomes': ['televisão', 'celular', 'notebook', 'geladeira', 'fogão'],
        'precos': [2000, 1500, 3500, 4000, 1500]}
```

Para acessar os valores, podemos usar um agrupamento de laços `for`:

```python
for chave, elementos in loja.items():
  print(f'Chave: {chave}\nElementos:')
  for dado in elementos:
    print(dado)
```

O primeiro laço, o mais externo, faz a leitura dos itens dentro do dicionário (chaves e elementos). Sabendo que os elementos são *listas*, podemos acessar os dados das listas com outro laço de repetição que está dentro do primeiro laço. O laço mais interno lê os elementos de cada lista por vez e imprime os valores dentro deles.

Além disso, podemos realizar operações comuns em listas, como: adicionar, remover ou contar itens na lista associada a uma chave do dicionário.

### Para saber mais: funções built-in

Durante as aulas, nós trabalhamos diretamente com várias funções built-in que são aquelas pré-definidas e disponíveis por padrão no Python. Elas funcionam como ferramentas úteis para realizar tarefas comuns, como conversões de tipo, operações matemáticas, manipulação de strings e muito mais, sem a necessidade de escrever código adicional.

Algumas funções built-in que já conhecemos são: `print()`, `input()`, `len()`, `int()`, `str()`, `float()`, `range()`, `chr()`, etc. Mas existem outras funções além dessas que também são muito úteis, como: `sum()`, `help()` e `dir()`. Vamos conhecê-las?

#### `sum()`

A função `sum()` permite somar os elementos de uma sequência ou estrutura de dados. No exemplo a seguir, vamos somar os preços de produtos:

```python
precos = [100.0, 400.0, 200.0]
soma = sum(precos)
soma
```
> 700.0


#### `help()`

A função `help()` é usada para acessar a documentação de funções, métodos e outros elementos do Python. Ela exibe informações em inglês sobre a funcionalidade, sintaxe e uso de um objeto específico. Para usar essa função, basta passar o elemento desejado entre parênteses. Por exemplo, vamos verificar a documentação da função `print`:

```python
help(print)
```
>  Help on built-in function print in module builtins:

> print(...)
>    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)

>    Prints the values to a stream, or to sys.stdout by default.
>    Optional keyword arguments:
>    file:  a file-like object (stream); defaults to the current sys.stdout.
>    sep:   string inserted between values, default a space.
>    end:   string appended after the last value, default a newline.
>    flush: whether to forcibly flush the stream.

#### `dir()`

Por fim, a função `dir()` é usada para exibir uma lista de atributos e métodos associados a um elemento. Por exemplo, vamos descobrir todos os atributos e métodos de uma lista:

```python
lista = [1, 2, 3]
dir(lista)
```
> `['__add__', '__class__', '__contains__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__iadd__', '__imul__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__rmul__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'append', 'clear', 'copy', 'count', 'extend', 'index', 'insert', 'pop', 'remove', 'reverse', 'sort']`

É interessante conhecer várias funções do built-in e como funcionam, pois elas auxiliam bastante na criação de códigos. Para saber mais sobre funções, você pode acessar a [documentação do Python](https://docs.python.org/3/library/functions.html).